### PR TITLE
Handle top-level sets while ignoring initProvider diffs.

### PR DESCRIPTION
**Pull request by mergenci**
### Description of your changes

* Handle top-level sets while eliminating initProvider exclusive diffs.
* Remove all length keys, which are suffixed with % or #, from initProvider exclusive diffs. For maps and sets, these length keys have no effect when applying the diff. For lists, length keys are required to apply diff.
* Nested sets to be tested.

I have:

- [x] Read and followed Upjet's [contribution process].
- [ ] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

Create an IAM Policy and get its ARN. Create an IAM Role and then edit it to add Policy ARN to `initProvider.managedPolicyArns`.

[contribution process]: https://github.com/crossplane/upjet/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: pkg/controller/external_nofork.go [hunk 1]
```diff
@@ -626,18 +627,17 @@ func getIgnoredFieldsArray(format string, forProvider, initProvider []any) []str
...
 	for i := range initProvider {
 		// Construct the full field path with array index and prefix.
-		fieldPath := fmt.Sprintf("%s[%d]", format, i)
+		fieldPath := fmt.Sprintf("%s.%d", format, i)
 		if i < len(forProvider) {
 			if _, ok := initProvider[i].(map[string]any); ok {
 				ignored = append(ignored, getIgnoredFieldsMap(fieldPath+".%s", forProvider[i].(map[string]any), initProvider[i].(map[string]any))...)
 			}
 			if _, ok := initProvider[i].([]any); ok {
-				ignored = append(ignored, getIgnoredFieldsArray(fieldPath+"%s", forProvider[i].([]any), initProvider[i].([]any))...)
```
Comment (ulucinar) [inline pkg/controller/external_nofork.go:679]: Looks like although the first argument is named as `format`, it's not actually a format parameter. Could you please open a bug regarding this?

File: pkg/controller/external_nofork.go [hunk 2]
```diff
@@ -267,44 +266,12 @@ func (c *NoForkConnector) Connect(ctx context.Context, mg xpresource.Managed) (m
...
 
-func deleteInstanceDiffAttribute(instanceDiff *tf.InstanceDiff, paramKey string) error {
-	delete(instanceDiff.Attributes, paramKey)
-
-	keyComponents := strings.Split(paramKey, ".")
-	if len(keyComponents) < 2 {
-		return nil
-	}
-
-	keyComponents[len(keyComponents)-1] = "%"
```
Comment (ulucinar) [inline pkg/controller/external_nofork.go:278]: So does it mean Terraform does not actually care about the length diffs if the remaning `ResourceAttrDiff` are properly calculated?

File: pkg/controller/external_nofork.go [hunk 3]
```diff
@@ -313,27 +280,61 @@ func filterInitExclusiveDiffs(tr resource.Terraformed, instanceDiff *tf.Instance
...
-			if err := deleteInstanceDiffAttribute(instanceDiff, keyToIgnore); err != nil {
-				return errors.Wrapf(err, "cannot delete key %q from instance diff", keyToIgnore)
-			}
+			delete(instanceDiff.Attributes, attributeKey)
 
-			keyComponents := strings.Split(keyToIgnore, ".")
+			// TODO: tags-tags_all implementation is AWS specific.
+			// Consider making this logic independent of provider.
+			keyComponents := strings.Split(attributeKey, ".")
 			if keyComponents[0] != "tags" {
```
Comment (ulucinar) [inline pkg/controller/external_nofork.go:298]: Yes, we will need to consider this also while initializing the `tags_all` argument. These have to be provider specific.

File: pkg/controller/external_nofork.go (hunk 1 above)
Comment (mergenci) [reply pkg/controller/external_nofork.go:679]: Done: [`ignore_changes` formatting is wrong for list of lists](https://github.com/crossplane/upjet/issues/295)

File: pkg/controller/external_nofork.go (hunk 2 above)
Comment (mergenci) [reply pkg/controller/external_nofork.go:278]: Yes, for maps and sets. No, for lists. Indicated in comments.

File: pkg/controller/external_nofork.go (hunk 3 above)
Comment (mergenci) [reply pkg/controller/external_nofork.go:298]: I copied to-do comment to all places where we duplicate the logic, to make sure that we don't miss any when addressing concerns.

**Review by ulucinar — APPROVED** [review]
Thanks @mergenci!

**Pull request merged**